src/models/curriculum/problem.ts

```diff
@@ -20,14 +22,39 @@ const ModernProblemModel = types
     ordinal: types.integer,
     title: types.string,
     subtitle: "",
-    sections: types.array(SectionModel),
+    loadedSections: types.array(SectionModel),
     supports: types.array(SupportModel),
     config: types.maybe(types.frozen<Partial<ProblemConfiguration>>())
   })
   .views(self => ({
     get fullTitle() {
       return `${self.title}${self.subtitle ? `: ${self.subtitle}` : ""}`;
     },
+
+    // In order to support shared models in the sections. Each section has to
+    // have its own MST environment to hold its document's sharedModelManager.
+    // So each section has to be its own tree and cannot be a child of the
+    // problem.
+    get sections() {
+      return self.loadedSections.map(section => {
+        const sharedModelManager = new SharedModelDocumentManager();
+        const environment: ITileEnvironment = {
+          sharedModelManager
+        };
+        const sectionSnapshot = getSnapshot(section);
+        // We have to make a copy of the sectionSnapshot because child models modify
+        // their snapshots in place during preProcessor calls. The objects from
+        // getSnapshot are readonly
+        const sectionCopy = SectionModel.create(JSON.parse(JSON.stringify(sectionSnapshot)), environment);
```

Comment:
I would use something like `lodash.cloneDeep()` here rather than `stringify`/`parse`, but I don't feel too strongly about it.